staging/src/github.com/openshift/oc/pkg/cli/version/version.go

```diff
@@ -1,25 +1,185 @@
-func NewCmdVersion(fullName string, versionInfo version.Info, out io.Writer) *cobra.Command {
+type VersionOptions struct {
+	ClientOnly bool
+	ServerOnly bool
+	All        bool
+	Output     string
+
+	discoveryClient discovery.CachedDiscoveryInterface
+	oClient         configv1client.ClusterOperatorsGetter
+
+	genericclioptions.IOStreams
+}
+
+func NewVersionOptions(ioStreams genericclioptions.IOStreams) *VersionOptions {
+	return &VersionOptions{
+		IOStreams: ioStreams,
+	}
+
+}
+
+func NewCmdVersion(fullName string, f cmdutil.Factory, ioStreams genericclioptions.IOStreams) *cobra.Command {
+	o := NewVersionOptions(ioStreams)
 	cmd := &cobra.Command{
-		Use:   "version",
-		Short: "Display version",
-		Long:  "Display version",
+		Use:     "version",
+		Short:   "Print the client and server version numbers",
+		Long:    "Print the client and server version numbers for the current context",
+		Example: fmt.Sprintf(versionExample, fullName),
 		Run: func(cmd *cobra.Command, args []string) {
-			fmt.Fprintf(out, "%s %v\n", fullName, versionInfo)
+			cmdutil.CheckErr(o.Complete(f, cmd))
+			cmdutil.CheckErr(o.Validate())
+			cmdutil.CheckErr(o.Run())
 		},
 	}
-
+	cmd.Flags().BoolVar(&o.ClientOnly, "client", o.ClientOnly, "Client version only (no server required).")
+	cmd.Flags().BoolVar(&o.ServerOnly, "server", o.ServerOnly, "Servers version only.")
+	cmd.Flags().BoolVar(&o.All, "all", o.All, "Print full version information.")
+	cmd.Flags().StringVarP(&o.Output, "output", "o", o.Output, "One of 'yaml' or 'json'.")
 	return cmd
 }
+
+func (o *VersionOptions) Complete(f cmdutil.Factory, cmd *cobra.Command) error {
+	var err error
+	o.discoveryClient, err = f.ToDiscoveryClient()
+	// if we had an empty rest.Config, continue and just print out client information.
+	// if we had an error other than being unable to build a rest.Config, fail.
+	if err != nil && !clientcmd.IsEmptyConfig(err) {
+		return err
+	}
+	clientConfig, err := f.ToRESTConfig()
+	if err != nil && !clientcmd.IsEmptyConfig(err) {
+		return err
+	}
+	o.oClient, err = configv1client.NewForConfig(clientConfig)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (o *VersionOptions) Validate() error {
+	if o.Output != "" && o.Output != "yaml" && o.Output != "json" {
+		return errors.New(`--output must be 'yaml' or 'json'`)
+	}
+
+	return nil
+}
+
+func (o *VersionOptions) Run() error {
+	var (
+		errs        []error
+		versionInfo Version
+	)
+
+	if !o.ServerOnly {
+		clientVersion := version.Get()
+		versionInfo.ClientVersion = &clientVersion
+	}
+
+	if !o.ClientOnly {
+		if o.discoveryClient != nil {
+			// Always request fresh data from the server
+			o.discoveryClient.Invalidate()
+			serverVersion, err := o.discoveryClient.ServerVersion()
+			errs = append(errs, err)
+			versionInfo.ServerVersion = serverVersion
+		}
+		if o.oClient != nil {
+			openshiftVersion, err := o.oClient.ClusterOperators().Get("openshift-apiserver", metav1.GetOptions{})
```

Comment:
@smarterclayton @deads2k any better way to get the openshift version?

Reply:
I had either this or from clusterversion/version.  @soltysh was using this, so I figured it was as good as anything : )

Reply:
We can improve this later IMO. As long as we have a good fallback behaviour if this code fails.